### DISCUSSION TITLE
Test try: import module / except ModuleNotFoundError:

### DISCUSSION THREAD

**Discussion by takos22** [pytest-dev/pytest#8605] [Q&A]
I want to have a code coverage of 100% for my module, so I need to test:
```py
try:
    import module
except ModuleNotFoundError:
    # do something
else:
    # do something else
```
How can I test both the `except` and the `else`? I need to make it raise a `ModuleNotFoundError` at one point then nothing in another test.

**Comment by portyu9** [comment]
Hi @takos22 

The cleanest approach is to move the optional import behind a function and patch the import operation rather than trying to manipulate Python's real installation state.

For example:

```python
# app.py
import importlib


def load_optional():
    try:
        module = importlib.import_module("optional_module")
    except ModuleNotFoundError:
        return "missing"

    return "available"
```

Then both branches are straightforward to test:

```python
def test_missing(monkeypatch):
    def missing(name):
        raise ModuleNotFoundError(name)

    monkeypatch.setattr(app.importlib, "import_module", missing)

    assert app.load_optional() == "missing"


def test_available(monkeypatch):
    fake_module = object()

    monkeypatch.setattr(
        app.importlib,
        "import_module",
        lambda name: fake_module,
    )

    assert app.load_optional() == "available"
```

If the import is executed once at module-import time, testing both branches requires manipulating `sys.modules` and reloading the target module, which is much more brittle. Refactoring the optional import behind a small function makes the behavior directly testable.

Please mark this answer as accepted if it helped, thank you.
